### PR TITLE
Feature/top bottom countries

### PR DESCRIPTION
**Pull request by bimal125**
- Addresses xxxxxx (eg: #1)
- Depends on xxxxxx (eg: #1)

## Changes

* Add top 5 and low 5 indicator values in overview page

Mention related users here if any.

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] n+1 queries
- [x] flake8 issues
- [x] `print`
- [x] typos
- [x] unwanted comments

## This PR contains valid:

- [ ] tests
- [ ] permission checks (tests here too)

### CODE REVIEW

**Review by Rup-Narayan-Rajbanshi — APPROVED** [review]
looks good.

File: apps/visualization/utils.py [hunk 1]
```diff
@@ -556,3 +564,79 @@ async def get_global_combined_indicators(filters):
...
+                Countries.objects.filter(
+                    iso3=OuterRef('iso3')
+                ).order_by().values('country_name')[:1]
+            ),
+        ).order_by('max_indicator_month')
+    if is_top:
+        qs = qs.order_by('-indicator_value')[:5]
+    else:
+        qs = qs.order_by('indicator_value')[:5]
+    return get_unique_countries_data(qs)
```
Comment (thenav56) [inline apps/visualization/utils.py]: Let's add some documentation here why we are using different tables as well.

```suggestion
@sync_to_async
def get_indicator_stats_latest(
    emergency,
    region,
    indicator_id,
    is_top=False,
):
    from .types import IndicatorLatestStatsType

    existing_iso3 = Countries.objects.values('iso3')

    def _clean_filters(filters):
        # Filter out None values
        return {
            k: v
            for k, v in filters.items()
            if v is not None
        }

    filters = _clean_filters({
        'region': region,
        'emergency': emergency,
    })
    if indicator_id:
        filters.update({
            'indicator_id': indicator_id,
        })
        qs = DataCountryLevelMostRecent.objects.filter(
            **filters,
            iso3__in=existing_iso3,
            category='Global',
            indicator_value__gt=0
        ).order_by().values('iso3').annotate(
            indicator_value=F('indicator_value'),
            max_indicator_month=Max('indicator_month'),
            format=F('format'),
            country_name=F('country_name'),
        ).order_by('-max_indicator_month', 'subvariable')

    else:
        qs = CountryEmergencyProfile.objects.filter(
            **filters,
            iso3__in=existing_iso3,
            context_indicator_value__gt=0
        ).order_by().values('iso3').annotate(
            indicator_value=F('context_indicator_value'),
            max_indicator_month=Max('context_date'),
            format=F('format'),
            country_name=Subquery(
                Countries.objects.filter(
                    iso3=OuterRef('iso3')
                ).order_by().values('country_name')[:1]
            ),
        ).order_by('max_indicator_month')

    if is_top:
        qs = qs.order_by('-indicator_value')[:5]
    else:
        qs = qs.order_by('indicator_value')[:5]

    return [
        IndicatorLatestStatsType(
            iso3=data['iso3'],
            indicator_value=data['indicator_value'],
            format=data['format'],
            country_name=data['country_name'],
        ) for data in qs.distinct('iso3')
    ]
```

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: apps/visualization/utils.py (hunk 1 above)
Comment (Rup-Narayan-Rajbanshi) [reply apps/visualization/utils.py]: return [
        IndicatorLatestStatsType(
            iso3=data['iso3'],
            indicator_value=data['indicator_value'],
            format=data['format'],
            country_name=data['country_name'],
        ) for data in qs.distinct('iso3')
    ]
@thenav56 this codes gives an error  "Cannot create distinct fields once a slice has been taken.". So I tried 
if is_top:
        qs = qs.distinct('iso3').order_by('-indicator_value')[:5]
    else:
        qs = qs.distinct('iso3').order_by('indicator_value')[:5]

This leads to error '"annotate() + distinct(fields) is not implemented."
So, may be we should use the previous logic for return value.

Comment (thenav56) [reply apps/visualization/utils.py]: We don't need the `.distinct('iso3')` as we are using **iso3** for **GROUP BY**. Let's remove that and it should work.

Comment (Rup-Narayan-Rajbanshi) [reply apps/visualization/utils.py]: @thenav56 Sir, Done !

File: apps/visualization/utils.py [hunk 2]
```diff
@@ -267,22 +274,26 @@ def get_unique_countries_data(qs):
...
     else:
         all_filters = {
             'emergency': emergency,
         }
         filters = {k: v for k, v in all_filters.items() if v is not None}
+        if region:
+            iso3_list = existing_iso3.filter(region=region).values_list('iso3', flat=True)
+            filters['iso3__in'] = iso3_list
+        else:
+            filters['iso3__in'] = existing_iso3
```
Comment (thenav56) [inline apps/visualization/utils.py]: We need iso3 value in both case right? Also, can we just filter out the existing_iso3? if yes let's change the existing_iso3 to countries_qs and then let's filter it as following.

```suggestion
        if region:
            countries_qs = existing_iso3.filter(region=region)
        filters['iso3__in'] = countries_qs.values('iso3')
```

File: apps/visualization/utils.py [hunk 3]
```diff
@@ -357,8 +368,12 @@ def format_indicator_value(iso3, qs_map):
...
```
Comment (thenav56) [inline apps/visualization/utils.py]: Same as before.

File: apps/visualization/utils.py [hunk 4]
```diff
@@ -556,3 +571,78 @@ async def get_global_combined_indicators(filters):
...
+        ).order_by('-max_indicator_month', 'subvariable')
+    else:
+        filters = _clean_filters({
+            'emergency': emergency,
+        })
+        if region:
+            iso3_list = existing_iso3.filter(region=region).values_list('iso3', flat=True)
+            filters['iso3__in'] = iso3_list
+        else:
+            filters['iso3__in'] = existing_iso3
```
Comment (thenav56) [inline apps/visualization/utils.py]: Same here.

File: apps/visualization/utils.py [hunk 5]
```diff
@@ -556,3 +571,78 @@ async def get_global_combined_indicators(filters):
...
+            if v is not None
+        }
+    if indicator_id:
+        filters = _clean_filters({
+            'region': region,
+            'emergency': emergency,
+        })
+        filters.update({
+            'indicator_id': indicator_id,
+        })
```
Comment (thenav56) [inline apps/visualization/utils.py]:
```suggestion
        filters = _clean_filters({
            'region': region,
            'emergency': emergency,
            'indicator_id': indicator_id,
        })
```

**Review by thenav56 — CHANGES_REQUESTED** [review]

**Review by thenav56 — APPROVED** [review]

**Pull request merged**